File: DesignPatterns/01_GOF23/DP_examples/Go/02_Structural/Facade/subsystem.go
package Facade

import "fmt"

// 形状接口
type IShape interface {
	Draw()
}

// 圆形实现
type Circle struct {
	radius int
}

// 创建圆形
func NewCircle(radius int) *Circle {
	return &Circle{
		radius: radius,
	}
}

// 绘制圆形
func (c *Circle) Draw() {
	fmt.Printf("Draw a circle with radius %d\n", c.radius)
}

// 矩形实现
type Rectangle struct {
	width  int
	height int
}

// 创建矩形
func NewRectangle(width, height int) *Rectangle {
	return &Rectangle{
		width:  width,
		height: height,
	}
}

// 绘制矩形
func (r *Rectangle) Draw() {
	fmt.Printf("Draw a rectangle with width %d and height %d\n", r.width, r.height)
}

// 正方形实现
type Square struct {
	side int
}

// 创建正方形
func NewSquare(side int) *Square {
	return &Square{
		side: side,
	}
}

// 绘制正方形
func (s *Square) Draw() {
	fmt.Printf("Draw a square with side %d\n", s.side)
}

File: 01_GOF23/DP_examples/C/02_Structural/Facade/subsystem.c
#include "subsystem.h"
#include <stdlib.h>

// 圆形绘制方法
static void Circle_draw(IShape* self) {
    Circle* circle = (Circle*)self;
    printf("Draw a circle with radius %d\n", circle->radius);
}
static void Circle_destroy(IShape* self) {
    Circle* circle = (Circle*)self;
    free(circle);
}

// 初始化圆形
Circle* Circle_create(int radius) {
    Circle* circle = (Circle*)malloc(sizeof(Circle));
    if (circle) {
        circle->base.draw = Circle_draw;
        circle->radius = radius;
        circle->base.destroy = Circle_destroy;
    }
    return circle;
}

// 矩形绘制方法
static void Rectangle_draw(IShape* self) {
    Rectangle* rectangle = (Rectangle*)self;
    printf("Draw a rectangle with width %d and height %d\n", rectangle->width, rectangle->height);
}
static void Rectangle_destroy(IShape* self) {
    Rectangle* rectangle = (Rectangle*)self;
    free(rectangle);
}

// 初始化矩形
Rectangle* Rectangle_create(int width, int height) {
    Rectangle* rectangle = (Rectangle*)malloc(sizeof(Rectangle));
    if (rectangle) {
        rectangle->base.draw = Rectangle_draw;
        rectangle->width = width;
        rectangle->height = height;
        rectangle->base.destroy = Rectangle_destroy;
    }
    return rectangle;
}

// 正方形绘制方法
static void Square_draw(IShape* self) {
    Square* square = (Square*)self;
    printf("Draw a square with side %d\n", square->side);
}
static void Square_destroy(IShape* self) {
    Square* square = (Square*)self;
    free(square);
}

// 初始化正方形
Square* Square_create(int side) {
    Square* square = (Square*)malloc(sizeof(Square));
    if (square) {
        square->base.draw = Square_draw;
        square->side = side;
        square->base.destroy = Square_destroy;
    }
    return square;
}


